crates/wasi/src/preview2/ip_name_lookup.rs

```diff
@@ -5,35 +5,73 @@ use crate::preview2::poll::{subscribe, Pollable, Subscribe};
 use crate::preview2::{spawn_blocking, AbortOnDropJoinHandle, SocketError, WasiView};
 use anyhow::Result;
 use std::mem;
-use std::net::{Ipv6Addr, ToSocketAddrs};
+use std::net::{IpAddr, Ipv6Addr, ToSocketAddrs};
 use std::pin::Pin;
 use std::str::FromStr;
 use std::vec;
 use wasmtime::component::Resource;
 
-use super::network::{from_ipv4_addr, from_ipv6_addr};
+use super::WasiCtx;
 
 pub enum ResolveAddressStream {
-    Waiting(AbortOnDropJoinHandle<Result<Vec<IpAddress>, SocketError>>),
-    Done(Result<vec::IntoIter<IpAddress>, SocketError>),
+    Waiting(AbortOnDropJoinHandle<std::io::Result<Vec<IpAddr>>>),
+    Done(std::io::Result<vec::IntoIter<IpAddr>>),
 }
 
-#[async_trait::async_trait]
-impl<T: WasiView> Host for T {
+/// A trait for providing a `IpNameLookup` implementation.
+pub trait WasiNetworkView: Send {
+    /// Given a name, resolve to a list of IP addresses
     fn resolve_addresses(
         &mut self,
-        network: Resource<Network>,
         name: String,
-    ) -> Result<Resource<ResolveAddressStream>, SocketError> {
-        let network = self.table().get(&network)?;
+    ) -> AbortOnDropJoinHandle<std::io::Result<Vec<IpAddr>>>;
+}
 
-        let host = parse(&name)?;
+/// The default implementation for `WasiIpNameLookupView`.
+#[derive(Debug, Clone, Default)]
+pub struct SystemNetwork {
+    allowed: bool,
+}
 
-        if !network.allow_ip_name_lookup {
-            return Err(ErrorCode::PermanentResolverFailure.into());
+impl SystemNetwork {
```

Comment:
Could this be removed in lieu of implementing `WasiNetworkView for WasiCtx` itself? (given that `WasiCtx` is needed to create this structure anyway it may not buy a whole lot)

Reply:
If a user wants to have some logic where they conditionally switch between either their own network implementation and `SystemNetwork`, they need some way of constructing the `SystemNetwork`. Without this constructor, they would not be able to do so. We could do the following: 
* make all fields of `SystemNetwork` public
* Make the constructor just take a simple `bool` instead of a `WasiCtx`
* Leave it as is

I think I'd be most in favor of the last two options.

Reply:
Hm I'm not sure I understand? Given `impl WasiNetworkView for WasiCtx` then if users are themselves implementing `WasiNetworkView` for their own custom type then conditionally using the default behavior would be calling `self.ctx.the_method_name()` so I'm not sure where the need for having this separate `SystemNetwork` structure arises?